### PR TITLE
Azure AMQP package

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Moving from Azure.Core dll to separate package.
Also a few API changes.

### CODE REVIEW

**Comment by jsquire** [comment]
We should probably consider adding a CODEOWNERS entry for this.   I'd like to be on the list, if you decide to do so.

File: sdk/core/Azure.Core.Amqp/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Azure Core AMQP shared client library for .NET
+
+Azure.Core.Amqp contains Amqp primitives.
```
Comment (jsquire) [inline sdk/core/Azure.Core.Amqp/README.md:3]: (extreme) nit:  AMQP

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusMessage.cs [hunk 2]
```diff
@@ -337,9 +337,17 @@ public DateTimeOffset ScheduledEnqueueTime
...
         /// Gets or sets the raw Amqp message data that will be transmitted over the wire.
         /// This can be used to enable scenarios that require setting AMQP header, footer, property, or annotation
-        /// data that is not exposed as top level properties in the ServiceBusMessage.
+        /// data that is not exposed as top level properties in the <see cref="ServiceBusMessage"/>.
         /// </summary>
-        public AmqpAnnotatedMessage AmqpMessage { get; set; }
+        internal AmqpAnnotatedMessage AmqpMessage { get; set; }
+
+        /// <summary>
+        /// Gets or sets the raw Amqp message data that will be transmitted over the wire.
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusMessage.cs:345]: nit:  The comment appears incorrect; the accessor is get-only.

**Review by jsquire — APPROVED** [review]
Looks like a good first cut.   We probably want to log some work to flesh out the doc comments on the AMQP types in the near future.

File: sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (pakrym) [inline sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:1]: Are we expecting to have extra dependencies in this package?

File: sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+    <EnableClientSdkAnalyzers>false</EnableClientSdkAnalyzers>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="$(AzureCoreSharedSources)Argument.cs" Link="SharedSource\Azure.Core\Argument.cs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\Azure.Core.Experimental\src\Azure.Core.Experimental.csproj" />
```
Comment (weshaggard) [inline sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:17]: Should this have a dependency on experimental? We need to ensure we don't end up taking a shipping GA dependency on this package.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:17]: This will be removed once we ship System.BinaryData.

File: sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:1]: Hmm most likely not. why?

Comment (pakrym) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:1]: Why split the package then?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:1]: Ah, this was the recommendation from @KrzysztofCwalina. He wanted a separate dll for this functionality.

Comment (pakrym) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:1]: To get what benefit?

File: sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj [hunk 5]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <Description>Azure AMQP types</Description>
+    <AssemblyTitle>Microsoft Azure AMQP</AssemblyTitle>
+    <Version>1.0.0-beta.1</Version>
+    <PackageTags>Microsoft Azure AMQP</PackageTags>
+    <Nullable>enable</Nullable>
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+    <EnableClientSdkAnalyzers>false</EnableClientSdkAnalyzers>
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:9]: this isn't needed - removed.

File: sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+        /// <summary>
+        /// Creates a new Data body <see cref="AmqpAnnotatedMessage"/>.
+        /// </summary>
+        /// <param name="dataBody">The data sections comprising the message body.
+        /// <seealso href="http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-data"/>
+        /// </param>
+#pragma warning disable CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
+        public AmqpAnnotatedMessage(IEnumerable<BinaryData> dataBody)
+#pragma warning restore CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
```
Comment (MiYanni) [inline sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: Why not declare the fields nullable (`?`) instead of ignoring the error? You have logic to check to see if they are null anyway in the property getters.

**Review by MiYanni — APPROVED** [review]

File: sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: Because it is not really ever going to be null. We initialize it in the getter.

Comment (pakrym) [reply sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: It is null until someone calls the getter.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: But why would users need to care about this?

Comment (pakrym) [reply sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: This warning is not for users but for you.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs:40]: My bad.. I thought the warning was about the public properties needing to be nullable, not the private fields.

File: sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <Description>Azure AMQP types</Description>
+    <AssemblyTitle>Microsoft Azure AMQP</AssemblyTitle>
+    <Version>1.0.0-beta.1</Version>
+    <PackageTags>Microsoft Azure AMQP</PackageTags>
+    <Nullable>enable</Nullable>
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+    <NoWarn>$(NoWarn);AZC0001;AZC0012</NoWarn>
```
Comment (pakrym) [inline sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:9]: Why did you need the AZC0012 suppression?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Amqp/src/Azure.Core.Amqp.csproj:9]: https://github.com/Azure/azure-sdk-for-net/pull/16341
